src/events/attributes.rs

```diff
@@ -333,187 +278,2076 @@ impl<'a> From<(&'a str, &'a str)> for Attribute<'a> {
+            //                  |  ^
+            //             offset
+            Some((offset, _)) => return self.skip_value(slice, offset),
+            // Input: `    key  =  `
+            //                  |  ^
+            //             offset
+            None => return None,
+        };
+
+        match iter.find(|(_, &b)| b == quote) {
+            // Input: `    key  =  "   "`
+            //                         ^
+            Some((e, b'"')) => Some(e),
+            // Input: `    key  =  '   '`
+            //                         ^
+            Some((e, _)) => Some(e),
+
+            // Input: `    key  =  "   `
+            // Input: `    key  =  '   `
+            //                         ^
+            // Closing quote not found
+            None => None,
         }
+    }
 
-        if len <= self.position {
-            return None;
+    #[inline]
+    fn check_for_duplicates(
+        &mut self,
+        slice: &[u8],
+        key: Range<usize>,
+    ) -> Result<Range<usize>, AttrError> {
+        if self.check_duplicates {
+            if let Some(prev) = self
+                .keys
+                .iter()
+                .find(|r| slice[(*r).clone()] == slice[key.clone()])
+            {
+                return Err(AttrError::Duplicated(key.start, prev.start));
+            }
+            self.keys.push(key.clone());
         }
+        Ok(key)
+    }
+
+    /// # Parameters
+    ///
+    /// - `slice`: content of the tag, used for checking for duplicates
+    /// - `key`: Range of key in slice, if iterator in HTML mode
+    /// - `offset`: Position of error if iterator in XML mode
+    #[inline]
+    fn key_only(&mut self, slice: &[u8], key: Range<usize>, offset: usize) -> Option<AttrResult> {
+        Some(if self.html {
+            self.check_for_duplicates(slice, key).map(Attr::Empty)
+        } else {
+            Err(AttrError::ExpectedEq(offset))
+        })
+    }
+
+    #[inline]
+    fn double_q(&mut self, key: Range<usize>, value: Range<usize>) -> Option<AttrResult> {
+        self.state = State::Next(value.end + 1); // +1 for `"`
+
+        Some(Ok(Attr::DoubleQ(key, value)))
+    }
+
+    #[inline]
+    fn single_q(&mut self, key: Range<usize>, value: Range<usize>) -> Option<AttrResult> {
+        self.state = State::Next(value.end + 1); // +1 for `'`
 
-        let mut bytes = self.bytes.iter().enumerate().skip(self.position);
+        Some(Ok(Attr::SingleQ(key, value)))
+    }
 
-        // key starts after the whitespace
-        let start_key = match bytes
-            .by_ref()
-            .skip_while(|&(_, &b)| !is_whitespace(b))
-            .find(|&(_, &b)| !is_whitespace(b))
-        {
-            Some((i, _)) => i,
-            None => return attr!(self.position..len),
+    pub fn next(&mut self, slice: &[u8]) -> Option<AttrResult> {
+        let mut iter = match self.recover(slice) {
+            Some(offset) => (offset..).zip(slice[offset..].iter()),
```

Comment:
Using this instead of 
```rust
slice.iter().enumerate().skip(offset)
```
really improves performance around by 10 us (~15%), according to Criterion, so applied it